README.md
# explain8601
Convert an ISO 8601 expression to a human readable description. ISO 8601-1:2019 and parts of ISO 8601-2:2019 are supported, older versions of the standard and RFC 3339 are not supported.

## Installation
You can download a release as a jar file from the GitHub releases page. Alternatively see the build section below.

## Usage
The basic usage is

    $ java -jar explain8601-0.1.0-standalone.jar [opts] -e expression

Use the ``-d`` option to show the intermediate results of the conversion.

## Build
Clojure and Leiningen are required to build a jar file, after cloning this repo run:
    
    $ lein uberjar

### GraalVM native image
Additionally building a native binary using GraalVM is possible:
    
    $ export GRAALVM_HOME=/usr/lib/jvm/java-17-graalvm/bin # this path depends on your GraalVM installation
    $ $GRAALVM_HOME/gu install native-image # install the native image component
    $ lein native-image

## Examples
The following is a list of generated descriptions:
- '2022-22' represents the summer of 2022.
- '2022/2023' represents an interval from the year 2022 to the year 2023.
- 'P1YP2W' represents a duration of 1 year and 2 weeks.
- 'R2/2022-123T00:00/PT1H' represents an interval from the 123rd day of the year 2022 at 0 minutes past 0 hours, lasting for a duration of 1 hour, repeating 2 times.
- '2022Y7M30~DT1S' represents the 30th day (approximate) of july of the year 2022 at 1 second past 0 hours.

## Bugs
Not all features from ISO 8601-2:2019 are supported, missing are e.g.:
- grouped units
- selection rules
- ranges (``..``)

Expressions are currently not validated, it is therefore easy to produce accepted but invalid expressions.

## Further information
https://en.wikipedia.org/wiki/ISO_8601

freely available related standards:
- https://www.loc.gov/standards/datetime/
- https://standards.calconnect.org/csd/cc-18011.html

## License
AGPLv3


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]

## [0.1.0]
### Added
Initial features


test/explain8601/core_test.clj
(ns explain8601.core-test
  (:require [clojure.test :refer :all]
            [explain8601.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/explain8601/parser_test.clj
(ns explain8601.parser-test
  (:require [clojure.test :refer :all]
            [explain8601.parser :as parser]))

(deftest parser-test
  (let [test-cases {"" '()
                    "00:00" '([:representation [:time [:hour "0" "0"] ":" [:minute "0" "0"]]])
                    "12:34,5" '([:representation [:time [:hour "1" "2"] ":" [:minute "3" "4"] [:fraction "," "5"]]])
                    "1234" '([:representation [:date [:calendar-date [:year "1" "2" "3" "4"]]]]
                             [:representation [:time [:hour "1" "2"] [:minute "3" "4"]]])
                    "2022-01-02" '([:representation [:date [:calendar-date [:year "2" "0" "2" "2"] "-" [:month "0" "1"] "-" [:day "0" "2"]]]])
                    "2022-100" '([:representation [:date [:ordinal-date [:year "2" "0" "2" "2"] "-" [:day-of-year "1" "0" "0"]]]])
                    "2022100" '([:representation [:date [:ordinal-date [:year "2" "0" "2" "2"] [:day-of-year "1" "0" "0"]]]])
                    "2022-W01-7" '([:representation [:date [:week-date [:year "2" "0" "2" "2"] "-" "W" [:week "0" "1"] "-" [:weekday "7"]]]])
                    "2022W017" '([:representation [:date [:week-date [:year "2" "0" "2" "2"] "W" [:week "0" "1"] [:weekday "7"]]]])
                    "2022-01-02T03:04" '([:representation [:date-time [:year "2" "0" "2" "2"] "-" [:month "0" "1"] "-" [:day "0" "2"]
                                                           "T" [:time [:hour "0" "3"] ":" [:minute "0" "4"]]]])
                    "PT01H" '([:representation [:duration "P" "T" [:hours "0" "1"] "H"]])
                    "P0001-02" '([:representation [:duration "P" [:calendar-date [:year "0" "0" "0" "1"] "-" [:month "0" "2"]]]])
                    "2022-01/02" '([:representation [:interval [:year "2" "0" "2" "2"] "-" [:month "0" "1"] [:solidus "/"] [:calendar-date-tail-month [:month "0" "2"]]]])
                    "R3/2022/P1Y" '([:representation [:repeating-interval "R" [:repetitions "3"] [:solidus "/"]
                                                      [:interval [:date [:calendar-date [:year "2" "0" "2" "2"]]] [:solidus "/"] [:duration "P" [:years "1"] "Y"]]]])}]
    (doall (map
            (fn [s]
              (testing s (is (=
                              (parser/parse-all-8601 s)
                              (test-cases s)))))
            (keys test-cases)))))

test/explain8601/explain8601_test.clj
(ns explain8601.explain8601-test
  (:require [clojure.test :refer :all]
            [explain8601.parser :as parser]
            [explain8601.transformer :as transformer]))

(deftest explain8601-test
  (let [explain8601 (fn [expression]
                      (-> expression
                          parser/parse-all-8601-1
                          parser/parse-all-8601-2
                          parser/parse-all-8601-3
                          transformer/transform-8601
                          (transformer/descriptions->description expression)))
        test-cases {"" "'' does not appear to be a valid expression.\n"
                    "P3W" "'P3W' represents a duration of 3 weeks.\n"
                    "-P1W" "'-P1W' represents a duration of 1 week in the reverse direction.\n"
                    "-1234S2" "'-1234S2' represents a year between -1200 and -1299, estimated to be -1234.\n"
                    "2022-123" "'2022-123' represents the 123rd day of the year 2022.\n"
                    "2022-W12-X" "'2022-W12-X' represents an unspecified day of the 12th week of the year 2022.\n"
                    "01:02+03" "'01:02+03' represents 2 minutes past 1 hours (3 hours ahead of UTC).\n"
                    "R3/P1Y/2022" "'R3/P1Y/2022' represents an interval lasting for a duration of 1 year, ending the year 2022, repeating 3 times.\n"}]
    (doall (map
            (fn [s]
              (testing s (is (=
                              (explain8601 s)
                              (test-cases s)))))
            (keys test-cases)))))

project.clj
(defproject explain8601 "0.1.0"
  :description "Convert an ISO 8601 expression to a human readable description"
  :url "https://github.com/dokutan/explain8601"
  :license {:name "AGPL-3.0"}
  :dependencies [[org.clojure/clojure "1.10.3"]
                 [org.clojure/tools.cli "1.0.206"]
                 [instaparse "1.4.12"]
                 [zprint "1.2.3"]]
  :plugins [[lein-kibit "0.1.8"]
            [lein-codox "0.10.8"]
            [io.taylorwood/lein-native-image "0.3.1"]]
  :native-image {:graal-bin :env/GRAALVM_HOME
                 :opts ["--verbose"
                        "--report-unsupported-elements-at-runtime"
                        "--initialize-at-build-time"
                        ]}
  :main ^:skip-aot explain8601.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all
                       :jvm-opts ["-Dclojure.compiler.direct-linking=true"]}
             :native-image {:jvm-opts ["-Dclojure.compiler.direct-linking=true"]}})


src/explain8601/core.clj
(ns explain8601.core
  (:require [explain8601.parser :as parser]
            [explain8601.transformer :as transformer]
            [zprint.core :as zp]
            [clojure.tools.cli :as cli]
            [clojure.string :as string])
  (:gen-class))

(def ^:private usage
  "explain8601 [opts] -e expression")

(def ^:private cli-options
  [["-e" "--expression EXPRESSION" "the expression that should be explained"]
   ["-d" "--debug" "show intermediate results for debugging"]
   ["-h" "--help" "show this help message"]])

(defn -main
  "The main function"
  [& args]
  (let [opts (cli/parse-opts args cli-options)]
    (cond
      (:errors opts)
      (println (string/join "\n" (:errors opts)))

      (or (:help (:options opts))
          (not (:expression (:options opts))))
      (println (str usage "\n" (:summary opts)))

      :else
      (let [expression (:expression (:options opts))
            parse-tree-1 (parser/parse-all-8601-1 expression)
            parse-tree-2 (parser/parse-all-8601-2 parse-tree-1)
            parse-tree-3 (parser/parse-all-8601-3 parse-tree-2)
            descriptions (transformer/transform-8601 parse-tree-3)
            description (transformer/descriptions->description descriptions expression)]
        (when (:debug (:options opts))
          (zp/czprint expression)
          (println "↓")
          (zp/czprint parse-tree-1)
          (println "↓")
          (zp/czprint parse-tree-2)
          (println "↓")
          (zp/czprint parse-tree-3)
          (println "↓")
          (zp/czprint descriptions)
          (println "↓"))
        (println (string/trim-newline description))))))


src/explain8601/parser.clj
(ns explain8601.parser
  (:require [explain8601.invalidator :as invalidator]
            [instaparse.core :as insta]
            [clojure.string :as string]
            [clojure.set :as set]))

(defn- triml0
  "Remove leading zeros from `s`, ignoring a leading minus."
  [s]
  (let [r (string/replace s #"^(-)?0*" "$1")]
    (cond
      (empty? r)
      "0"

      (or (string/starts-with? r ".") (string/starts-with? r ","))
      (str "0" r)

      :else
      r)))

(def ^:private flattenv (comp vec flatten))

(def ^:private parse-8601
  (insta/parser
   (clojure.java.io/resource "8601.abnf")
   :output-format :hiccup ; :hiccup or :enlive
   :input-format :abnf))

(defn- parse-date-year
  "Parse a :date-year node"
  [& c]
  (let [sign (first c)
        c (if (vector? sign)
            (rest c)
            c)
        sign (if (and (vector? sign)
                      (= :minus (first sign)))
               "-"
               "")
        exponent (first (filter vector? c))
        exponent (if (and (vector? exponent)
                          (= :exponent (first exponent)))
                   (Integer/parseInt (second exponent))
                   0)
        sig-digits (last (filter vector? c))
        sig-digits (when (and (vector? sig-digits)
                              (= :significant-digits (first sig-digits)))
                     (Integer/parseInt (second sig-digits)))
        year (string/join (filter string? c))
        year (str year (.repeat "0" exponent))
        r (if sig-digits
            {:year (str sign year)
             :min (str sign
                       (.substring year 0 sig-digits)
                       (.repeat "0" (- (.length year) sig-digits)))
             :max (str sign
                       (.substring year 0 sig-digits)
                       (.repeat "9" (- (.length year) sig-digits)))}
            {:year year})]
    [:date-year r]))

(defn- parse-qualifier
  "Parse a :qualifier node"
  [qualifier]
  (condp = qualifier
    "?" [:qualifier #{:uncertain}]
    "~" [:qualifier #{:approximate}]
    "%" [:qualifier #{:uncertain :approximate}]
    [:qualifier #{}]))

(defn- unit->duration
  "Convert keyword keys in `m` to their plural form"
  [m]
  (merge
   (dissoc
    m
    :year
    :month
    :week
    :day-of
    :day
    :hour
    :minute
    :second)
   (condp (fn [key coll] (contains? coll key)) m
     :year {:years (get m :year)}
     :month {:months (get m :month)}
     :week {:weeks (get m :week)}
     :day-of-year {:days (get m :day-of-year)}
     :day {:days (get m :day)}
     :hour {:hours (get m :hour)}
     :minute {:minutes (get m :minute)}
     :second {:seconds (get m :second)}
     {})))

(defn- parse-duration
  "Parse a :duration node"
  [& c]
  (let [minus (first c)
        c (if (and (vector? minus)
                   (= :minus (first minus)))
            (rest c)
            c)
        c (map unit->duration (filter map? (flattenv c)))]
    [:duration {:components (vec c)
                :reverse (vector? minus)}]))

(defn- parse-interval-limit
  "Parse the start or end of an interval"
  [limit type-hint]
  (let [value (cond
                (and (vector? limit)
                     (= :interval-limit-unknown (ffirst limit)))
                [:unknown]

                (and (vector? limit)
                     (= :interval-limit-open (ffirst limit)))
                [:open]

                (and (= :calendar-date (ffirst limit))
                     (or (= :time (first (second limit)))
                         (pos? (count (filter map? limit)))))
                (flattenv [:calendar-date-time (filter map? (flatten limit))])

                (and (= :week-date (ffirst limit))
                     (or (= :time (first (second limit)))
                         (pos? (count (filter map? limit)))))
                (flattenv [:week-date-time (filter map? (flatten limit))])

                (and (= :ordinal-date (ffirst limit))
                     (or (= :time (first (second limit)))
                         (pos? (count (filter map? limit)))))
                (flattenv [:ordinal-date-time (filter map? (flatten limit))])

                (or (= :calendar-date-tail (ffirst limit))
                    (= :week-date-tail (ffirst limit))
                    (= :ordinal-date-tail (ffirst limit)))
                (flattenv [(filter map? (flatten limit))])

                :else
                (flattenv limit))]
    {(if (= :duration (first value))
       :duration
       type-hint)
     value}))

(defn- copy-missing-values
  "Copy missing values in an interval end from the interval start"
  [start end]
  (if (and (get end :end)
           (get start :start)
           (or (map? (first (get end :end)))
               (.contains (str (first (get end :end))) "-tail")))
    (let [end (get end :end)
          start (get start :start)
          end-keys (apply merge (map (fn [e] {(set (keys e)) e}) (filter map? end)))
          start-keys (apply merge (map (fn [e] {(set (keys e)) e}) (filter map? start)))]
      {:end (vec (concat [(first start)] (vals (merge start-keys end-keys))))})
    end))

(defn- parse-interval
  "Parse a :duration node"
  [& c]
  (loop [start []
         end []
         state :start
         c c]
    (cond
      (zero? (count c))
      (let [start (parse-interval-limit start :start)
            end (parse-interval-limit end :end)
            end (copy-missing-values start end)]
        [:interval
         (merge start end)])

      (and (vector? (first c)) (= :solidus (ffirst c)))
      (recur
       start
       end
       :end
       (rest c))

      (= :start state)
      (recur
       (conj start (first c))
       end
       state
       (rest c))

      :else
      (recur
       start
       (conj end (first c))
       state
       (rest c)))))

(defn- parse-recurring-interval
  "Parse a :recurring-interval node"
  [& c]
  (let [interval (second c)
        repetitions (first c)
        repetitions (if
                     (or (= 1 (count repetitions))
                         (= "-1" (second repetitions)))
                      :unbounded
                      (second repetitions))]
    [(first interval)
     (assoc (second interval)
            :repetitions repetitions)]))

(defn- parse-date-time
  "Flatten a date and time expression"
  [kw]
  (fn [& c]
    (flattenv [kw (filter map? (flattenv c))])))

(defn- parse-timezone
  "Parse a :timezone node"
  [& c]
  (cond
    (and (= 1 (count c)) (=  "Z" (first c)))
    [:timezone {:timezone :utc}]

    :else
    [:timezone
     {:timezone
      (if (and (= :pm (ffirst (filter vector? c)))
               (= "+" (second (first (filter vector? c)))))
        :plus
        :minus)
      :offset (vec (filter map? c))}]))

(defn- parse-component
  "Generic function for parsing components like year, month, …"
  [kw]
  (fn [& c]
    (let [c (map
             (fn [n]
               (cond
                 (and (vector? n) (= :minus (first n)))
                 "-"
                 (and (vector? n) (= :pm (first n)))
                 (if (= "+" (second n))
                   ""
                   "-")
                 :else
                 n))
             c)
          qualifier (second (first (filter vector? c)))
          value (triml0 (.replace (string/join (filter string? c)) "." ","))]
      (if qualifier
        {kw value
         :qualifier qualifier}
        {kw value}))))

(defn- add-start-end
  "Add :start and :end values"
  [k digits]
  (fn [m]
    (merge
     m
     {:start (str (get m k) (.repeat "0" digits))
      :end   (str (get m k) (.repeat "9" digits))})))

(defn- move-qualifier
  "Apply free qualifiers to components"
  [kw]
  (fn [& c]
    (loop [c (reverse c)
           qualifier nil
           return []]
      (if (pos? (count c))
        (if (and (vector? (first c))
                 (= :qualifier (ffirst c)))
          (recur (rest c)
                 (set/union qualifier (second (first c)))
                 return)
          (recur (rest c)
                 qualifier
                 (conj
                  return
                  (if (nil? qualifier)
                    (first c)
                    (update-in (first c) [:qualifier] set/union qualifier)))))
        (vec (reverse (conj return kw)))))))

(defn month->grouping
  "Convert a month to a grouping if the month exceeds 12"
  [m]
  (try
    (if (> (Integer/parseInt (:month m)) 12)
      {:grouping (:month m)}
      m)
    (catch Exception _ m)))

(defn parse-all-8601-1
  "The first pass of the parser, accepts a string `s`"
  [s]
  (if (invalidator/invalid-string? s)
    '()
    (distinct (insta/parses parse-8601 s))))

(defn parse-all-8601-2
  "The second pass of the parser"
  [tree]
  (distinct
   (insta/transform
    {:expression identity
     :DIGIT identity
     :DIGITX identity
     :unspecified-value identity

     :date-year parse-date-year
     :qualifier parse-qualifier
     :duration parse-duration

     :century (comp (add-start-end :century 2) (parse-component :century))
     :century-e (comp (add-start-end :century 2) (parse-component :century))
     :century-expanded (comp (add-start-end :century 2) (parse-component :century))
     :decade (comp (add-start-end :decade 1) (parse-component :decade))
     :decade-e (comp (add-start-end :decade 1) (parse-component :decade))
     :decade-expanded (comp (add-start-end :decade 1) (parse-component :decade))
     :year (parse-component :year)
     :year-e (parse-component :year)
     :year-expanded (parse-component :year)
     :month (comp month->grouping (parse-component :month))
     :month-e (comp month->grouping (parse-component :month))
     :grouping (parse-component :grouping)
     :grouping-e (parse-component :grouping)
     :week (parse-component :week)
     :week-e (parse-component :week)
     :day (parse-component :day)
     :day-e (parse-component :day)
     :weekday (parse-component :weekday)
     :weekday-e (parse-component :weekday)
     :day-of-year (parse-component :day-of-year)
     :day-of-year-e (parse-component :day-of-year)
     :hour (parse-component :hour)
     :hour-fraction (parse-component :hour)
     :hour-e (parse-component :hour)
     :hour-e-fraction (parse-component :hour)
     :minute (parse-component :minute)
     :minute-fraction (parse-component :minute)
     :minute-e (parse-component :minute)
     :minute-e-fraction (parse-component :minute)
     :second (parse-component :second)
     :second-fraction (parse-component :second)
     :second-e (parse-component :second)
     :second-e-fraction (parse-component :second)
     :years (parse-component :years)
     :months (parse-component :months)
     :weeks (parse-component :weeks)
     :days (parse-component :days)
     :hours (parse-component :hours)
     :minutes (parse-component :minutes)
     :seconds (parse-component :seconds)

     :timezone parse-timezone
     :timezone-e parse-timezone

     :calendar-date (move-qualifier :calendar-date)
     :calendar-date-e (move-qualifier :calendar-date)
     :calendar-date-day (move-qualifier :calendar-date)
     :calendar-date-month (move-qualifier :calendar-date)
     :calendar-date-grouping (move-qualifier :calendar-date)
     :calendar-date-year (move-qualifier :calendar-date)
     :calendar-date-decade (move-qualifier :calendar-date)
     :calendar-date-century (move-qualifier :calendar-date)
     :calendar-date-day-e (move-qualifier :calendar-date)
     :calendar-date-month-e (move-qualifier :calendar-date)
     :calendar-date-grouping-e (move-qualifier :calendar-date)
     :calendar-date-year-e (move-qualifier :calendar-date)
     :calendar-date-decade-e (move-qualifier :calendar-date)
     :calendar-date-century-e (move-qualifier :calendar-date)
     :week-date (move-qualifier :week-date)
     :week-date-e (move-qualifier :week-date)
     :week-date-day (move-qualifier :week-date)
     :week-date-week (move-qualifier :week-date)
     :week-date-day-e (move-qualifier :week-date)
     :week-date-week-e (move-qualifier :week-date)
     :ordinal-date (move-qualifier :ordinal-date)
     :ordinal-date-e (move-qualifier :ordinal-date)
     :calendar-date-time-e (move-qualifier :calendar-date-time)
     :week-date-time-e (move-qualifier :week-date-time)
     :ordinal-date-time-e (move-qualifier :ordinal-date-time)
     :calendar-date-time (move-qualifier :calendar-date-time)
     :week-date-time (move-qualifier :week-date-time)
     :ordinal-date-time (move-qualifier :ordinal-date-time)
     :time (move-qualifier :time)
     :time-e (move-qualifier :time)
     :time-hour (move-qualifier :time)
     :time-hour-e (move-qualifier :time)
     :time-minute (move-qualifier :time)
     :time-minute-e (move-qualifier :time)
     :time-second (move-qualifier :time)
     :time-second-e (move-qualifier :time)}
    tree)))

(defn parse-all-8601-3
  "The third pass of the parser"
  [tree]
  (let [result (distinct
                (insta/transform
                 {:interval parse-interval
                  :interval-e parse-interval
                  :recurring-interval parse-recurring-interval

                  :time (parse-date-time :time)
                  :calendar-date (parse-date-time :calendar-date)
                  :week-date (parse-date-time :week-date)
                  :ordinal-date (parse-date-time :ordinal-date)
                  :calendar-date-time (parse-date-time :calendar-date-time)
                  :week-date-time (parse-date-time :week-date-time)
                  :ordinal-date-time (parse-date-time :ordinal-date-time)
                  :date-time identity
                  :date-time-e identity
                  :date identity
                  :date-e identity}
                 tree))
        result-no-time (filter (fn [e] (not= :time (first e))) result)]
    (if (and
         (> (count result) 1)
         (pos? (count result-no-time)))
      result-no-time
      result)))

src/explain8601/transformer.clj
(ns explain8601.transformer
  (:require [instaparse.core :as insta]
            [clojure.string :as string]))

(declare transform-8601)

(defn- find-map-with
  "Finds the first map in `s` that has key `k`"
  [s k]
  (first
   (filter
    (fn [m]
      (and (map? m)
           (get m k)))
    s)))

(defn- join-and
  "Join items using ', ' and ' and '"
  [items]
  (let [start (reverse (rest (reverse items)))
        end (last items)]
    (if (empty? start)
      (str end)
      (str
       (string/join ", " start)
       " and "
       end))))

(defn- singular
  "Remove trailing 's' from `str` if `i` = 1"
  [str i]
  (try
    (if (or (> (Float/parseFloat (.replace i "," ".")) 1)
            (< (Float/parseFloat (.replace i "," ".")) -1)
            (= i "0"))
      str
      (subs str 0 (dec (count str))))
    (catch Exception _ str)))

(defn- parenthesize
  "Add parantheses aroiund `s` if `s` is not empty"
  [s]
  (if (= "" s)
    s
    (str " (" s ")")))

(defn- transform-qualifier
  "Describe the `qualifiers`"
  [qualifiers]
  (cond
    (and (contains? qualifiers :uncertain)
         (contains? qualifiers :approximate))
    " (uncertain and approximate)"

    (contains? qualifiers :uncertain)
    " (uncertain)"

    (contains? qualifiers :approximate)
    " (approximate)"

    :else
    ""))

(defn- transform-duration-component
  "Describes the component `c` of a duration"
  [c]
  (reduce-kv
   (fn [_ k v]
     (.replace (str v " " (singular (str k) v) (transform-qualifier (get c :qualifier))) ":" ""))
   ""
   (dissoc c :qualifier)))

(defn- transform-date-year
  "Tranform a :date-year node"
  [props]
  (if (get props :min)
    (format
     "a year between %s and %s, estimated to be %s"
     (get props :min)
     (get props :max)
     (get props :year))
    (format "the year %s" (get props :year))))

(defn- grouping->name
  "Convert a :grouping value to its description"
  [grouping]
  (case grouping
    "21" "spring"
    "22" "summer"
    "23" "autumn"
    "24" "winter"
    "25" "spring (northern hemisphere)"
    "26" "summer (northern hemisphere)"
    "27" "autumn (northern hemisphere)"
    "28" "winter (northern hemisphere)"
    "29" "spring (southern hemisphere)"
    "30" "summer (southern hemisphere)"
    "31" "autumn (southern hemisphere)"
    "32" "winter (southern hemisphere)"
    "33" "first quarter"
    "34" "second quarter"
    "35" "third quarter"
    "36" "fourth quarter"
    "37" "first quadrimester"
    "38" "second quadrimester"
    "39" "third quadrimester"
    "40" "first semester"
    "41" "second semester"
    (str "grouping " grouping)))

(defn- weekday->name
  "Convert the number of a weekday to its name"
  [weekday]
  (case weekday
    "1" "monday"
    "2" "tuesday"
    "3" "wednesday"
    "4" "thursday"
    "5" "friday"
    "6" "saturday"
    "7" "sunday"
    "X" "an unspecified day"
    "X*" "an unspecified day"
    (str "day " weekday)))

(defn- month->name
  "Convert the number of a month to its name"
  [month]
  (case month
    "1" "january"
    "2" "february"
    "3" "march"
    "4" "april"
    "5" "may"
    "6" "june"
    "7" "july"
    "8" "august"
    "9" "september"
    "10" "october"
    "11" "november"
    "12" "december"
    "XX" "an unspecified month"
    "X*" "an unspecified month"
    (str "month " month)))

(defn- ordinal
  "Return the ordinal version of `i`"
  [i]
  (condp (fn [x y] (string/ends-with? y x)) i
    "11" (str i "th")
    "12" (str i "th")
    "13" (str i "th")
    "1" (str i "st")
    "2" (str i "nd")
    "3" (str i "rd")
    (str i "th")))

(defn- describe-timezone
  "Describe a timezone"
  [& props]
  (let [props (flatten props)
        timezone (find-map-with props :timezone)
        hour (find-map-with (:offset timezone) :hour)
        minute (find-map-with (:offset timezone) :minute)
        second (find-map-with (:offset timezone) :second)
        timezone (:timezone timezone)
        offset (join-and
                (filter
                 string?
                 [(when hour
                    (str
                     (:hour hour)
                     (singular " hours" (:hour hour))
                     (transform-qualifier (:qualifier hour))))
                  (when minute
                    (str
                     (:minute minute)
                     (singular " minutes" (:minute minute))
                     (transform-qualifier (:qualifier minute))))
                  (when second
                    (str
                     (:second second)
                     (singular " seconds" (:second second))
                     (transform-qualifier (:qualifier second))))]))]
    (case timezone
      :utc
      "UTC"

      :plus
      (str offset " ahead of UTC")

      :minus
      (str offset " behind UTC")

      "")))

(defn- transform-timezone
  "Transform a :timezone node"
  [props & _]
  (str
   "the time zone "
   (describe-timezone props)))

(defn- transform-calendar-date
  "Transform a :calendar-date node"
  [& props]
  (cond
    (:century (first props))
    (format
     "the century from %s to %s%s%s"
     (:start (first props))
     (:end (first props))
     (transform-qualifier (:qualifier (first props)))
     (parenthesize (describe-timezone props)))

    (:decade (first props))
    (format
     "the decade from %s to %s%s%s"
     (:start (first props))
     (:end (first props))
     (transform-qualifier (:qualifier (first props)))
     (parenthesize (describe-timezone props)))

    (and (:year (first props))
         (:grouping (second props)))
    (format
     "the %s%s of %s%s%s"
     (grouping->name (:grouping (second props)))
     (transform-qualifier (:qualifier (second props)))
     (:year (first props))
     (transform-qualifier (:qualifier (first props)))
     (parenthesize (describe-timezone props)))


    :else
    (let [year (find-map-with props :year)
          month (find-map-with props :month)
          day (find-map-with props :day)]
      (str
       (if day
         (str "the " (ordinal (:day day)) " day" (transform-qualifier (:qualifier day)) " of ")
         "")
       (if month
         (str (month->name (:month month)) (transform-qualifier (:qualifier month)) " of ")
         "") 
       "the year " (:year year) (transform-qualifier (:qualifier year))
       (parenthesize (describe-timezone props))))))

(defn- transform-week-date
  "Transform a :week-date node"
  [& props]
  (let [year (find-map-with props :year)
        week (find-map-with props :week)
        weekday (find-map-with props :weekday)]
    (str
     (if weekday
       (str (weekday->name (:weekday weekday)) (transform-qualifier (:qualifier weekday)) " of ")
       "")
     (if week
       (str "the " (ordinal (:week week)) " week" (transform-qualifier (:qualifier week)) " of ")
       "")
     "the year " (:year year) (transform-qualifier (:qualifier year))
     (parenthesize (describe-timezone props)))))

(defn- transform-ordinal-date
  "Transform a :ordinal-date node"
  [& props]
  (let [year (find-map-with props :year) 
        day (find-map-with props :day-of-year)]
    (str
     (if day
       (str "the " (ordinal (:day-of-year day)) " day" (transform-qualifier (:qualifier day)) " of ")
       "")
     "the year " (:year year) (transform-qualifier (:qualifier year))
     (parenthesize (describe-timezone props)))))

(defn- transform-time
  "Transform a :time node"
  [& props]
  (let [hour (find-map-with props :hour)
        minute (find-map-with props :minute)
        second (find-map-with props :second)]
    (str
     (join-and
      (filter
       string?
       [(when minute
          (str (:minute minute) (singular " minutes" (:minute minute)) (transform-qualifier (:qualifier minute))))
        (when second
          (str (:second second) (singular " seconds" (:second second)) (transform-qualifier (:qualifier second))))]))
     (if (or minute second) " past " "")
     (if hour
       (:hour hour)
       "0")
     " hours" (transform-qualifier (:qualifier hour))
     (parenthesize (describe-timezone props)))))

(defn- transform-duration
  "Tranform a :duration node"
  [props & _]
  (format
   "a duration of %s%s"
   (join-and
    (mapv transform-duration-component (get props :components)))
   (if (get props :reverse)
     " in the reverse direction"
     "")))

(defn- transform-interval-repetitions
  "Describe the number of repetitions of an interval"
  [repetitions]
  (case repetitions
    nil ""
    "0" ", not repeating"
    "1" ", repeating once"
    :unbounded ", repeating an unbounded number of times"
    (format ", repeating %s times" repetitions)))

(defn- transform-interval
  "Transform an :interval node"
  [props & _]
  (cond
    (and (:start props) (:end props) (= 1 (count (:start props))))
    (format
     "an interval with %s start, ending %s%s"
     (.replace (str (first (:start props))) ":" "")
     (first (transform-8601 (seq [(:end props)])))
     (transform-interval-repetitions (:repetitions props)))

    (and (:start props) (:end props) (= 1 (count (:end props))))
    (format
     "an interval starting %s, end %s%s"
     (first (transform-8601 (seq [(:start props)])))
     (.replace (str (first (:end props))) ":" "")
     (transform-interval-repetitions (:repetitions props)))

    (and (:start props) (:end props))
    (format
     "an interval from %s to %s%s"
     (first (transform-8601 (seq [(:start props)])))
     (first (transform-8601 (seq [(:end props)])))
     (transform-interval-repetitions (:repetitions props)))

    (and (:start props) (:duration props))
    (format
     "an interval from %s, lasting for %s%s"
     (first (transform-8601 (seq [(:start props)])))
     (first (transform-8601 (seq [(:duration props)])))
     (transform-interval-repetitions (:repetitions props)))

    (and (:end props) (:duration props))
    (format
     "an interval lasting for %s, ending %s%s"
     (first (transform-8601 (seq [(:duration props)])))
     (first (transform-8601 (seq [(:end props)])))
     (transform-interval-repetitions (:repetitions props)))
    
    :else
    (str "an interval " props)))

(defn- transform-set
  "Transform a :set node"
  [& elements]
  (str
   "the set containing:\n"
   (string/join (map (fn [s] (str "  ◦ " s "\n")) elements))))

(defn- transform-set-single
  "Transform a :set-single node"
  [& elements]
  (str
   "a single element of the set containing:\n"
   (string/join (map (fn [s] (str "  ◦ " s "\n")) elements))))

(defn transform-8601
  [tree]
  (insta/transform
   {:date-year transform-date-year
    :calendar-date transform-calendar-date
    :week-date transform-week-date
    :ordinal-date transform-ordinal-date
    :time transform-time
    :timezone transform-timezone
    :calendar-date-time (fn [& p] (str (apply transform-calendar-date p) " at " (apply transform-time p)))
    :week-date-time (fn [& p] (str (apply transform-week-date p) " at " (apply transform-time p)))
    :ordinal-date-time (fn [& p] (str (apply transform-ordinal-date p) " at " (apply transform-time p)))
    :interval transform-interval
    :duration transform-duration
    :set transform-set
    :set-single transform-set-single}
   tree))

(defn descriptions->description
  "Convert a sequence of descriptions in to a single string"
  [descriptions expression]
  (.replace
   (.replace
    (case (count descriptions)
      0
      (format
       "'%s' does not appear to be a valid expression.\n"
       expression)
      1
      (format
       "'%s' represents %s.\n"
       expression
       (first descriptions))
      (format
       "'%s' is ambiguous and could represent one of the following:\n%s"
       expression
       (string/join (map (fn [s] (str " • " s "\n")) descriptions))))
    "\n.\n" "\n")
   "\n\n" "\n"))